### DISCUSSION TITLE
Firmware Updater (Beta)

### DISCUSSION THREAD

**Discussion by Ancyker** [sanni/cartreader#910] [General Chat]
The firmware updater is ready for testing. For now, it only works if your OSCR has firmware with `ENABLE_UPDATER` already set.

**Comment by Ancyker** [comment]
The Firmware Updater now allows you to update to master. The server checks for updates to master every 30 minutes. The hash displayed after master, i.e. `master-90fa37c491494bd42c1476660713d3ce1ffc7a80`, is the hash of the latest commit it has. 

When an update is made to master any old binaries for it are automatically deleted. Because of that, choosing master will generally make it take longer to flash since the server is less likely to have a cached version of the firmware and will need to compile it first.

But this should be useful when people are posting issues and asking for help but aren't running the latest development version. Hopefully being able to use the updater to flash master will make it easier for people. They can also use it to switch back to one of the release versions after testing (if they want to).

**Comment by riri9292** [comment]
After updating from v13.1 to v13.5 on the firmware update page, nothing appears on the cartreader screen. How can I get the screen to work again?
When I try to run arduino.exe with administrator privileges to see what I can do, I just get the warning in the image and can't do anything.
Windows10

Warning:
This application was configured to use a bundled Java Runtime Environment but the runtime is missing or corrupted.

**Reply by sanni** [reply]
You need to unpack the zip file first.

**Reply by riri9292** [reply]
The zip file has already been unzipped.

**Reply by riri9292** [reply]
I re-downloaded the zip file and tried "Updating Manually" and it solved the problem. It appears that the file was not downloaded properly.

**Comment by Ancyker** [comment]
The firmware updater now sets the RTC if the option is selected.

**Comment by Ancyker** [comment]
Added support for 115200 baud. It attempts 9600 first, and if it doesn't get the expected reply, it then tries 115200.

<img width="540" height="471" alt="image" src="https://github.com/user-attachments/assets/dd92d7d0-4c12-41dd-8650-c7b39db2beb2" />

This is primarily to add the foundation to support updating when using the ANSI serial option of the upcoming firmware revamp.

![OSCR being controlled using a PuTTY serial terminal application](https://github.com/user-attachments/assets/d81d979c-5cf6-49a9-913b-b566301971f4)

I'm also planning on making a way to control the OSCR from the browser as well, but that will be something to explain elsewhere.

**Comment by dreimer1986** [comment]
Could you please paste the config.h used for the base firmware you are building there? I tried to replicate it in the official tools, but I always get a different looking result or an error that my files are too big to flash...

**Reply by Ancyker** [reply]
It uses the defaults from `Config.h`, only defining the hardware version and any requested flags. It also always enables the config file.
